### ISSUE TITLE
Record user last connection timestamp

### ISSUE THREAD

**Issue by fvanderbiest** [georchestra/georchestra#2271]
This information could be displayed in the console.
This is useful to some platform admins.

We currently record their use of webservices, not their latest CAS login action.

**Comment by fvanderbiest** [comment]
This would be a "nice to have" for the Geo2France project, ping @bchartier

**Comment by fvanderbiest** [comment]
Also interested: GeoGrandEst

**Comment by MaelREBOUX** [comment]
Me ! Me ! Me ! Me !
Since 2016 :p

**Comment by pmauduit** [comment]
Since the talk took place on IRC, there are several technical ways of achieving this:

* Using the slapo-lastbind overlay (https://manpages.ubuntu.com/manpages/bionic/man5/slapo-lastbind.5.html) :
  * :+1: :-1: :question: this will also track users authenticated via basic-auth
  * :-1: it won't work for setups making use of something else than the geOrchestra OpenLDAP (wink wink @MaelREBOUX)
* logging onto cas via log4j2 config (suggested by @jusabatier):
  * :+1: :-1: :question: will only track users authenticating via CAS (web-based)

**Comment by jusabatier** [comment]
- logging onto Security Proxy via log4j2 config : 
  - :+1: this will also track users authenticated via basic-auth
  - :-1: this will generate more logs (each request) / :+1: those logs can be used to generate metrics

**Comment by landryb** [comment]
i thought a bit about the pros/cons of enabling the lastbind overlay, was a bit concerned at first that it might generate *tons* of writes to the ldap server (eg if you have many connected users using an authenticated WMS...) **but** after looking at the code a bit, i'm relieved to say that it has been thought ahead and is handled by a configurable timeout/precision, per comment in https://github.com/openldap/openldap/blob/master/contrib/slapd-modules/lastbind/lastbind.c#L25

it has been recently changed in https://github.com/openldap/openldap/commit/7a4b42600b4fc149e1d61784fa19a63be63412dc (so for openldap 2.6.1, per https://lists.openldap.org/hyperkitty/list/openldap-announce@openldap.org/thread/VGSH3HMNASR6PP72CHV5BVNID5WOUXUN/) , previously there was a dedicated `olcLastBindPrecision` attribute in the overlay to configure the precision, but now apparently it's supposed to be "the backend's precision", whatever that is.. i'll dig further.

so i think in the coming days i'll have a look at:
- enabling the lastbind overlay by default, with a conservative precision (say 1h) - in the ansible playbook and the georchestra repo
- doing a PR on the console code to display the `authTimestamp` attribute value on a user page, if the attribute is set/exists (default to 'never' or 'unknown' otherwise ?).

I know it'll only work for instances using openldap, but my opinion is that it's a nice thing to have, somewhat easy to do, and most instances should be on openldap (and iirc RM will migrate to it ;)) ... dunno if 389ds or the microsoft thing have an equivalent attribute/overlay though.

**Comment by pmauduit** [comment]
> dunno if 389ds or the microsoft thing have an equivalent attribute/overlay though.

tested on an Active Directory, it seems that there is an attribute, but no idea in how it is expressed:

```
lastLogon: 133198048656546857
```

Maybe something related to https://devblogs.microsoft.com/oldnewthing/20090306-00/?p=18913

**Comment by jusabatier** [comment]
Thanks @landryb  for those investigation.

I don't know if it's possible, but it would be also usefull to have the account creation date if we want to use it for inactive users cleanup. So we can ignore the recently created account for which it should be normal to never connect.

**Comment by landryb** [comment]
> Thanks @landryb for those investigation.
> 
> I don't know if it's possible, but it would be also usefull to have the account creation date if we want to use it for inactive users cleanup. So we can ignore the recently created account for which it should be normal to never connect.

well for that technically there's an operational default attribute in openldap:
```
$ldapsearch -x -b dc=georchestra,dc=org uid=testuser createTimestamp|grep ^create
createTimestamp: 20201201143347Z
```
but that's openldap-specific - and there's also `modifyTimestamp:` for the last modification

**Comment by landryb** [comment]
ok, after some testing it requires to load the `ppolicy` module too, otherwise slapd complains:
```
Feb  2 10:58:51 demo slapd[330]: slap_global_control: unrecognized control: 1.3.6.1.4.1.42.2.27.8.5.1
```

so what i have for now:
```
$cat lastbind.ldif
dn: cn=module,cn=config
objectClass: olcModuleList
cn: module
olcModulePath: /usr/lib/ldap
olcModuleLoad: lastbind
olcModuleLoad: ppolicy

dn: olcOverlay=lastbind,olcDatabase={2}mdb,cn=config
objectClass: olcLastBindConfig
objectClass: olcOverlayConfig
objectClass: olcConfig
objectClass: top
olcOverlay: lastbind
olcLastBindPrecision: 3600
$ sudo ldapadd -Y EXTERNAL -H ldapi:/// -f lastbind.ldif
```
and after a `testuser`  connection, i have the new `authTimestamp` attribute updated:
```
$ldapsearch -x -b dc=georchestra,dc=org uid=testuser authTimestamp|egrep '^(dn|auth)'
dn: uid=testuser,ou=users,dc=georchestra,dc=org
authTimestamp: 20230202100501Z
```

**Comment by landryb** [comment]
i have something that 'almost works'

![image](https://user-images.githubusercontent.com/3421760/216307839-c9a3b91c-c9e0-48de-9176-bfabf9903311.png)

but the console backend doesn't seem to query the LDAP operational attribute straight away, which is expected - so the value is empty. I'll dig further.

**Comment by pierrejego** [comment]
You should modify here to add the new attribute
https://github.com/georchestra/georchestra/blob/master/ldap-account-management/src/main/java/org/georchestra/ds/users/AccountDaoImpl.java#L542

**Comment by landryb** [comment]
> You should modify here to add the new attribute https://github.com/georchestra/georchestra/blob/master/ldap-account-management/src/main/java/org/georchestra/ds/users/AccountDaoImpl.java#L542

i already have a local chunk/diff in this area, but `authTimestamp` is an operational attribute, and as such iirc client code doesnt have such attributes when requesting an entry, it needs to specify explicitely the requested attribute. That's how it works for `ldapsearch`, you dont get operational attributes by default...

i also need to figure out the right field type, since it's a timestamp with a Z at the end, different from the type used for `shadowExpire` and `privacyPolicyAgreementDate`

**Comment by landryb** [comment]
![image](https://user-images.githubusercontent.com/3421760/216562764-2622d3d1-e6a0-403e-bb81-c03bccf7b006.png)

that's what i have with my wip. good enough ? i'm reluctant to add more datetime/display formatting.. dunno "where" is the best place to do that.

**Comment by pmauduit** [comment]
I'm wondering if it still works on a LDAP without the extra overlays. I guess it should be ok to query extra fields which don't exist, so I'd expect it to work as well with the added code.

**Comment by landryb** [comment]
> I'm wondering if it still works on a LDAP without the extra overlays. I guess it should be ok to query extra fields which don't exist, so I'd expect it to work as well with the added code.

`authTimestamp` attr doesnt exist for `testReviewer` in my dev setup, since that user didnt login since i enabled the overlay, and the code handles that fine (eg parse only if not null) - IIRC ldap protocol gracefully handles being asked for non-existent attributes.

**Comment by landryb** [comment]
implemented/fixed by #3889
